### PR TITLE
Add compass reader to the project

### PR DESCRIPTION
**Pull request by fabien0102**
### Summary

First version of the compass reader! Every detail how to test should be on the README.md 😉

### CODE REVIEW

**Comment by fabien0102** [comment]
Bien vu pour la doc ! 👍 

Oui, pour le moment, l’idée c’est de rester le plus simple possible, du coup si le serveur tourne, la caméra tourne et lis les valeurs 😉 Faut encore que j’fasse des tests un peu plus poussé, avec différents tags sur la table toussa toussa pour voir si on a besoin de plus

File: compass_reader/compass.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+
+            # Read markers
+            paramaters = aruco.DetectorParameters_create()
+            corners, ids, rejectedImgPoints = aruco.detectMarkers(
+                gray, ARUCO_DICT, parameters=paramaters)
+
+            # Save the value if the compass is visible
+            try:
+                if isinstance(ids, np.ndarray):
```
[the remark this answers is not in the document]
Comment (clemsciences) [reply compass_reader/compass.py]: It's okay to use isinstance in this case

[the remark this answers is not in the document]
Comment (sylvaing19) [reply compass_reader/compass.py]: I'm just talking about the unnecessary `if` statement

File: compass_reader/compass.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                        c = corners[compass[0][0]][0]
+                        c0 = c[0]
+                        c1 = c[1]
+                        delta = [c1[0] - c0[0], c1[1] - c0[0]]
+                        angle_in_rad = math.atan2(delta[0], delta[1])
+
+                        self.values += "N" if angle_in_rad > 0 else "S"
+
+                        # Cut the values to have the last results only
+                        self.values = self.values[BUFFER_SIZE * -1:]
```
[the remark this answers is not in the document]
Comment (clemsciences) [reply compass_reader/compass.py]: I like the second suggestion

File: compass_reader/compass.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                if isinstance(ids, np.ndarray):
+                    compass = np.where(ids == COMPASS_ID)
+                    if len(compass) > 0:
+                        c = corners[compass[0][0]][0]
+                        c0 = c[0]
+                        c1 = c[1]
+                        delta = [c1[0] - c0[0], c1[1] - c0[0]]
+                        angle_in_rad = math.atan2(delta[0], delta[1])
+
+                        self.values += "N" if angle_in_rad > 0 else "S"
```
[the remark this answers is not in the document]
Comment (fabien0102) [reply compass_reader/compass.py]: Also not really fan, but I'm trying to be a real python developer 😅

File: compass_reader/compass.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+                        # Cut the values to have the last results only
+                        self.values = self.values[BUFFER_SIZE * -1:]
+                        
+                        if (DEBUG):
```
[the remark this answers is not in the document]
Comment (fabien0102) [reply compass_reader/compass.py]: For the record, pylint sucks! 😅 (or I don't know how to configure it 🙄 )

File: compass_reader/compass.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                            print(self.values)
+
+                        time.sleep(.5)
+            except:
```
[the remark this answers is not in the document]
Comment (fabien0102) [reply compass_reader/compass.py]: I'm a prudent person ^^ We never know what can happen 😛

File: compass_reader/compass.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+            # Save the value if the compass is visible
+            try:
+                if isinstance(ids, np.ndarray):
+                    compass = np.where(ids == COMPASS_ID)
+                    if len(compass) > 0:
+                        c = corners[compass[0][0]][0]
+                        c0 = c[0]
+                        c1 = c[1]
+                        delta = [c1[0] - c0[0], c1[1] - c0[0]]
```
[the remark this answers is not in the document]
Comment (fabien0102) [reply compass_reader/compass.py]: mood 😅 and every other coordinate is a tuple, so… `[x, y]` seams appropriate

File: README.md [hunk 7]
```diff
@@ -76,3 +76,30 @@ Et de compiler grâce à maven:
     mvn clean compile assembly:single
 
 Le fichier généré est `chariot-hl.jar` placé dans `high_level/target`. Des scripts placés dans `high_level` permet de l'utiliser facilement.
+
```
[the remark this answers is not in the document]
Comment (PFGimenez) [reply README.md]: Plutôt `pip3` que `pip` non ?

[the remark this answers is not in the document]
Comment (fabien0102) [reply README.md]: Ça dépends comment t’as installé ton python 😉 (chez moi `pip` pointe sur python 3)

[the remark this answers is not in the document]
Comment (fabien0102) [reply README.md]: image

Ça fait pas une grande différence ^^

File: compass_reader/compass.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+        # Aruco config
+        self._aruco_dict = aruco.Dictionary_get(aruco.DICT_4X4_250)
+        self._aruco_params = aruco.DetectorParameters_create()
+
+        # Start `run` in a separate thread
+        thread = threading.Thread(target=self.run)
+        thread.daemon = True
+        thread.start()
+        self.should_stop = False
```
Comment (PFGimenez) [inline compass_reader/compass.py]: `self.should_stop = False` devrait être réalisé avant `thread.start()` car le thread vérifie la valeur de `self.should_stop`

File: compass_reader/compass.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PFGimenez) [reply compass_reader/compass.py]: You should not use `assert` on expected values nor catch `AssertionError`. Assert is a fail-fast protection that should never fail in runtime. This is why it's `AssertionError` and not `AssertionException`.

File: compass_reader/compass.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+WEBCAM_ID = 0
+DEBUG = True
+
+class CompassWatcher:
+    def __init__(self):
+        # Error
+        self.error = None
+
+        # Compass values
+        self.values = ""
```
Comment (PFGimenez) [inline compass_reader/compass.py]: Why storing the values in a string instead of a list?

Comment (fabien0102) [reply compass_reader/compass.py]: It's easier to compare and print for debug 😉 (I'm lazy)

Comment (PFGimenez) [reply compass_reader/compass.py]: As long as it works…

Comment (fabien0102) [reply compass_reader/compass.py]: Works perfectly! It's to permit this -> `if self.values == "S" * BUFFER_SIZE:`

Comment (fabien0102) [reply compass_reader/compass.py]: Otherwise I need a loop… it's boring ^^

Comment (PFGimenez) [reply compass_reader/compass.py]: Yes, I saw this horror… You are now a python master, I guess!

Comment (fabien0102) [reply compass_reader/compass.py]: I still prefer the JS version ❤️ 
```js
"S".repeat(bufferSize)
```

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (fabien0102) [reply README.md]: Va donc pour `pip3` 👍

File: compass_reader/compass.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (PFGimenez) [reply compass_reader/compass.py]: That's best to explicitly check the exceptions. Your code will for example catch an `KeyboardException` (aka Ctrl+C), and that's not something you want.

[the remark this answers is not in the document]
Comment (fabien0102) [reply compass_reader/compass.py]: I know, I know, I just need to figure what kind of expections I can have there (I'm still learning 😅 )

[the remark this answers is not in the document]
Comment (fabien0102) [reply compass_reader/compass.py]: The thing I need to catch (the unsafe part) is this -> `c = corners[compass[0][0]][0]`

You can maybe help me a bit 😃 I could also check the shape, but I don't know how (yet)

File: compass_reader/compass.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        # Start `run` in a separate thread
+        self.should_stop = False
+        thread = threading.Thread(target=self.run)
+        thread.daemon = True
+        thread.start()
+
+    def run(self):
+        cap = cv2.VideoCapture(WEBCAM_ID)
+
+        while(self.should_stop == False):
```
Comment (clemsciences) [inline compass_reader/compass.py]: `while(not self.should_stop)` is prettier here.

Comment (fabien0102) [reply compass_reader/compass.py]: I've tried with `while(!self.should_stop)` 😅 thanks!

File: compass_reader/compass.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (PFGimenez) [reply compass_reader/compass.py]: You can check the shape with `if compass.shape == (2,2)`.

[the remark this answers is not in the document]
Comment (PFGimenez) [reply compass_reader/compass.py]: And if you want to know what exceptions you need to catch, just run your code without except and try to provoke one. It is usually sufficient.

[the remark this answers is not in the document]
Comment (fabien0102) [reply compass_reader/compass.py]: `AttributeError: 'tuple' object has no attribute 'shape'` Not convince yet about this `shape` checking 😅

[the remark this answers is not in the document]
Comment (PFGimenez) [reply compass_reader/compass.py]: My bad, I though it was a numpy array. You can use `len(compass)` to check its length in the first dimension. You should also check the length of its elements. For example: `if len(compass)==2 and len(compass[0])==2 and len(compass[1])==2`.

**Pull request merged**